File: implementations/python/src/foilbench_py/core/solver_validation.py
"""Shared validation for transactional solver entry points."""

import math

import numpy as np

from foilbench_py.core.geometry import NacaFoil
from foilbench_py.core.models import (
    CanonicalFlowState,
    ControlState,
    NumericalFailure,
    RestartState,
    Scenario,
)


def precision_tolerance(precision: str, *values: float) -> float:
    base = 1.0e-6 if precision == "float32" else 1.0e-12
    return base * max(1.0, *(abs(value) for value in values))


def validate_advance_request(
    current_time: float,
    control: ControlState,
    target_dt: float,
    precision: str,
) -> None:
    if not math.isfinite(target_dt) or target_dt <= 0.0:
        raise ValueError("target_dt must be finite and positive")
    if not all(
        math.isfinite(value)
        for value in (
            control.time,
            control.angle_degrees,
            control.angular_velocity_degrees,
        )
    ):
        raise NumericalFailure(
            "time_contract_failure",
            "control state must be finite",
            "time-mapping",
        )
    expected = current_time + target_dt
    tolerance = precision_tolerance(precision, current_time, expected, control.time)
    if abs(control.time - expected) > tolerance:
        raise NumericalFailure(
            "time_contract_failure",
            "control completion time disagrees with the requested interval",
            "time-mapping",
            {
                "expected_time": expected,
                "control_time": control.time,
                "target_dt": target_dt,
            },
        )


def validate_restart_state(start: RestartState) -> None:
    if not all(
        math.isfinite(value)
        for value in (start.time, start.angle_degrees, start.reynolds)
    ):
        raise ValueError("restart state must be finite")
    if start.time < 0.0 or start.reynolds <= 0.0:
        raise ValueError("restart time must be nonnegative and Reynolds positive")


def validate_canonical_import(
    state: CanonicalFlowState,
    scenario: Scenario,
    control: ControlState,
) -> None:
    if state.schema_version != 1:
        raise NumericalFailure(
            "incompatible_domain",
            "canonical schema version is unsupported",
            "canonical-import",
        )
    if state.dimension != scenario.domain.dimension:
        raise NumericalFailure(
            "incompatible_domain",
            "canonical dimension does not match the scenario",
            "canonical-import",
        )
    if state.resolution != scenario.domain.resolution:
        raise NumericalFailure(
            "incompatible_domain",
            "canonical resolution does not match the scenario",
            "canonical-import",
        )
    if state.periodic_axes != scenario.domain.periodic_axes:
        raise NumericalFailure(
            "incompatible_domain",
            "canonical periodic axes do not match the scenario",
            "canonical-import",
        )
    if state.precision != scenario.precision:
        raise NumericalFailure(
            "incompatible_domain",
            "canonical precision does not match the scenario",
            "canonical-import",
        )
    expected_z = 1 if state.dimension == 2 else state.resolution[2]
    expected_velocity_shape = (
        expected_z,
        state.resolution[1],
        state.resolution[0],
        state.dimension,
    )
    expected_density_shape = expected_velocity_shape[:-1]
    expected_dtype = np.dtype(state.precision)
    if (
        state.velocity.shape != expected_velocity_shape
        or state.velocity.dtype != expected_dtype
        or (
            state.density is not None
            and (
                state.density.shape != expected_density_shape
                or state.density.dtype != expected_dtype
            )
        )
    ):
        raise NumericalFailure(
            "incompatible_domain",
            "canonical array shape or dtype does not match metadata",
            "canonical-import",
        )
    if not np.isfinite(state.velocity).all() or (
        state.density is not None and not np.isfinite(state.density).all()
    ):
        raise NumericalFailure(
            "nonfinite_state",
            "canonical arrays must be finite at import time",
            "canonical-import",
        )
    state_bounds = np.asarray(state.bounds, dtype=np.float64)
    scenario_bounds = np.asarray(scenario.domain.bounds, dtype=np.float64)
    tolerance = precision_tolerance(
        scenario.precision,
        *state_bounds.reshape(-1).tolist(),
        *scenario_bounds.reshape(-1).tolist(),
    )
    if not np.allclose(state_bounds, scenario_bounds, rtol=0.0, atol=tolerance):
        raise NumericalFailure(
            "incompatible_domain",
            "canonical bounds do not match the scenario",
            "canonical-import",
        )
    if not all(
        math.isfinite(value)
        for value in (
            control.time,
            control.angle_degrees,
            control.angular_velocity_degrees,
        )
    ):
        raise NumericalFailure(
            "time_contract_failure",
            "import control must be finite",
            "canonical-import",
        )
    control_tolerance = precision_tolerance(
        scenario.precision,
        state.time,
        control.time,
        state.angle_degrees,
        control.angle_degrees,
        state.angular_velocity_degrees,
        control.angular_velocity_degrees,
    )
    if (
        abs(state.time - control.time) > control_tolerance
        or abs(state.angle_degrees - control.angle_degrees) > control_tolerance
        or abs(state.angular_velocity_degrees - control.angular_velocity_degrees)
        > control_tolerance
    ):
        raise NumericalFailure(
            "time_contract_failure",
            "canonical time or foil control disagrees with the import control",
            "canonical-import",
            {
                "state_time": state.time,
                "control_time": control.time,
                "state_angle": state.angle_degrees,
                "control_angle": control.angle_degrees,
            },
        )
    if state.dimension == 2:
        solid = NacaFoil(scenario.foil).mask(
            scenario.domain,
            control.angle_degrees,
        )
        if np.any(state.velocity[0][solid] != 0.0):
            nonzero_solid_cells = int(
                np.count_nonzero(
                    np.any(state.velocity[0][solid] != 0.0, axis=1)
                )
            )
            raise NumericalFailure(
                "postcondition_failure",
                "canonical solid-cell velocity must be exactly zero",
                "canonical-import",
                {"nonzero_solid_cells": nonzero_solid_cells},
            )

File: implementations/julia/src/solver_validation.jl
function precision_tolerance(::Type{T}, values::Real...) where {T<:AbstractFloat}
    base = T === Float32 ? 1.0e-6 : 1.0e-12
    scale = isempty(values) ? 1.0 : max(1.0, maximum(abs, values))
    return base * scale
end

function validate_advance_request(
    current_time::T,
    control::ControlState,
    target_dt::Real,
) where {T<:AbstractFloat}
    isfinite(target_dt) && target_dt > 0 ||
        throw(ArgumentError("target_dt must be finite and positive"))
    all(isfinite, (control.time, control.angle_degrees, control.angular_velocity_degrees)) ||
        throw(NumericalFailure(
            :time_contract_failure,
            "control state must be finite",
            Symbol("time-mapping"),
        ))
    target = T(target_dt)
    expected = current_time + target
    tolerance = precision_tolerance(T, current_time, expected, control.time)
    abs(control.time - expected) <= tolerance || throw(NumericalFailure(
        :time_contract_failure,
        "control completion time disagrees with the requested interval",
        Symbol("time-mapping"),
        Dict{String,Any}(
            "expected_time" => expected,
            "control_time" => control.time,
            "target_dt" => target,
        ),
    ))
    return target
end

function validate_restart_state(start::RestartState)
    all(isfinite, (start.time, start.angle_degrees, start.reynolds)) ||
        throw(ArgumentError("restart state must be finite"))
    start.time >= 0 || throw(ArgumentError("restart time must be nonnegative"))
    start.reynolds > 0 || throw(ArgumentError("restart Reynolds number must be positive"))
    return nothing
end

function validate_canonical_import(
    state::CanonicalFlowState{StateD,S},
    scenario::Scenario{ScenarioD,T},
    control::ControlState,
) where {StateD,ScenarioD,S,T<:AbstractFloat}
    state.schema_version in (1, 2) || throw(NumericalFailure(
        :incompatible_domain,
        "canonical schema version is unsupported",
        Symbol("canonical-import"),
    ))
    StateD == ScenarioD || throw(NumericalFailure(
        :incompatible_domain,
        "canonical dimension does not match the scenario",
        Symbol("canonical-import"),
    ))
    state.resolution == scenario.domain.resolution || throw(NumericalFailure(
        :incompatible_domain,
        "canonical resolution does not match the scenario",
        Symbol("canonical-import"),
    ))
    state.periodic_axes == scenario.domain.periodic_axes || throw(NumericalFailure(
        :incompatible_domain,
        "canonical periodic axes do not match the scenario",
        Symbol("canonical-import"),
    ))
    S === T || throw(NumericalFailure(
        :incompatible_domain,
        "canonical precision does not match the scenario",
        Symbol("canonical-import"),
    ))
    expected_z = StateD == 2 ? 1 : state.resolution[3]
    expected_velocity = (expected_z, state.resolution[2], state.resolution[1], StateD)
    size(state.velocity) == expected_velocity &&
        (state.density === nothing || size(state.density) == expected_velocity[1:3]) ||
        throw(NumericalFailure(
            :incompatible_domain,
            "canonical array shape does not match metadata",
            Symbol("canonical-import"),
        ))
    all(isfinite, state.velocity) &&
        (state.density === nothing || all(isfinite, state.density)) ||
        throw(NumericalFailure(
            :nonfinite_state,
            "canonical arrays must be finite at import time",
            Symbol("canonical-import"),
        ))
    tolerance = precision_tolerance(T, Iterators.flatten(state.bounds)..., Iterators.flatten(scenario.domain.bounds)...)
    all(
        abs(state.bounds[axis][side] - scenario.domain.bounds[axis][side]) <= tolerance
        for axis in 1:StateD for side in 1:2
    ) || throw(NumericalFailure(
        :incompatible_domain,
        "canonical bounds do not match the scenario",
        Symbol("canonical-import"),
    ))
    all(isfinite, (control.time, control.angle_degrees, control.angular_velocity_degrees)) ||
        throw(NumericalFailure(
            :time_contract_failure,
            "import control must be finite",
            Symbol("canonical-import"),
        ))
    control_tolerance = precision_tolerance(
        T,
        state.time,
        control.time,
        state.angle_degrees,
        control.angle_degrees,
        state.angular_velocity_degrees,
        control.angular_velocity_degrees,
    )
    (
        abs(state.time - control.time) <= control_tolerance &&
        abs(state.angle_degrees - control.angle_degrees) <= control_tolerance &&
        abs(state.angular_velocity_degrees - control.angular_velocity_degrees) <= control_tolerance
    ) || throw(NumericalFailure(
        :time_contract_failure,
        "canonical time or foil control disagrees with the import control",
        Symbol("canonical-import"),
        Dict{String,Any}(
            "state_time" => state.time,
            "control_time" => control.time,
            "state_angle" => state.angle_degrees,
            "control_angle" => control.angle_degrees,
        ),
    ))
    state.schema_version == 1 || state.geometry == scenario.foil || throw(NumericalFailure(
        :incompatible_geometry,
        "canonical geometry does not match the scenario",
        Symbol("canonical-import"),
    ))
    if StateD == 2
        geometry = NacaFoil(scenario.foil)
        solid = solid_mask(geometry, scenario.domain, control.angle_degrees)
        nonzero = 0
        for j in 1:state.resolution[2], i in 1:state.resolution[1]
            solid[i, j] || continue
            any(state.velocity[1, j, i, component] != zero(S) for component in 1:StateD) &&
                (nonzero += 1)
        end
        nonzero == 0 || throw(NumericalFailure(
            :postcondition_failure,
            "canonical solid-cell velocity must be exactly zero",
            Symbol("canonical-import"),
            Dict{String,Any}("nonzero_solid_cells" => nonzero),
        ))
    end
    return nothing
end
